sylt-compiler/src/typechecker.rs

```diff
@@ -923,7 +923,44 @@ impl<'c> TypeChecker<'c> {
                         "Only boolean expressions are valid if-statement conditions"
                     )
                 }
-                self.statement(pass)?;
+
+                fn null_check(ctx: &mut TypeChecker, condition: &Expression) -> Result<(Type, usize), ()> {
+                    match &condition.kind {
+                        ExpressionKind::Comparison(lhs, ComparisonKind::Is, rhs) => {
+                            let lhs_type = match &lhs.kind {
+                                ExpressionKind::TypeConstant(ty) =>
+                                    ctx.compiler.resolve_type(ty, ctx.compiler_context()),
+                                _ => return Err(()),
+                            };
+
+                            let rhs_name = match &rhs.kind {
+                                ExpressionKind::Get(ass) => {
+                                    ctx.assignable(ass, ctx.namespace).map_err(|_| ())?
+                                }
+                                _ => return Err(()),
+                            };
+
+                            if let Lookup::Value { slot: Some(slot), .. } = rhs_name {
+                                Ok((lhs_type, slot))
+                            } else {
+                                Err(())
+                            }
+                        }
+                        _ => Err(()),
+                    }
+                }
+
+                let check = null_check(self, condition);
+                if let Ok((ty, slot)) = check {
+                    let old_ty = self.stack[slot].ty.clone();
+                    self.stack[slot].ty = ty;
```

Comment:
Should'nt the null check create a new variable with the same name and a different type, sort of like shadowing, instead of mutating the stack? That would allow the variable to be changed inside a if statement (for example) without requiring the variable struct to be changed(?).

Reply:
This should simplify the implementation as well... I can't think of a counter argument, so I should try it!

Reply:
But this entire solution - of changing the lookup, is kinda limited and won't work for blob-fields. I think we as a whole want a different solution where we store extra information that we've deduced from the code, so I don't know how long this code will actually stay.

Reply:
Third times the charm - I don't know how useful it would be to not modify the variable, since this code would have to be changed to do a more `proper version` anyways. There are already a ton of problems with this. It doesn't handle:
 - globals
 - blob-fields
 - extenensions

So I've change my mind for the third time in 9 minutes. It's a good idea, but the code won't stay for long enough for the change to be worth it.

Reply:
I mean, this is not something that is very easy to do right. Not many languages have this feature except maybe TypeScript. It might be worth looking at what they do to solve this.